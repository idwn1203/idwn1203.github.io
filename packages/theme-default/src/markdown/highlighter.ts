import type { BundledLanguage } from 'shiki/langs'
import type { BundledTheme, CodeToHastOptions } from 'shiki'
import { getHighlighter } from 'shiki'
import type { Highlighter } from '@sveltepress/vite'
import { LRUCache } from 'lru-cache'
import { createTransformerFactory } from '@shikijs/twoslash'
import { createTwoslasher, rendererFloatingSvelte } from '@sveltepress/twoslash'
import { themeOptionsRef } from '../index.js'
import { processCommands } from './commands.js'

const DEFAULT_SUPPORT_LANGUAGES: BundledLanguage[] = ['svelte', 'sh', 'js', 'html', 'ts', 'md', 'css', 'scss']

const cache = new LRUCache<string, any>({ max: 1024 })

async function createHighlighterWithThemeAndLangs() {
  const highlighterConfig = themeOptionsRef.value?.highlighter
  const langs = highlighterConfig?.languages || DEFAULT_SUPPORT_LANGUAGES
  const darkTheme = highlighterConfig?.themeDark ?? 'night-owl'
  const lightTheme = highlighterConfig?.themeLight ?? 'vitesse-light'
  const shikiHighlighter = await getHighlighter({
    themes: [darkTheme, lightTheme],
    langs,
  })

  const _highlighter: Highlighter = async (code, lang) => {
    const shikiOptions: CodeToHastOptions<BundledLanguage, BundledTheme> = {
      lang,
      themes: {
        dark: darkTheme,
        light: lightTheme,
      },
      transformers: [],
    }
    if (themeOptionsRef.value?.highlighter?.twoslash) {
      shikiOptions.transformers?.push(createTransformerFactory(await createTwoslasher())({
        langs: ['ts', 'tsx', 'svelte'],
        renderer: rendererFloatingSvelte({
          lang: 'ts',
        }),
      }))
    }
    return shikiHighlighter.codeToHtml(code, shikiOptions)
      .replace(/\{/g, '&#123;')
      .replace(/\}/g, '&#125;')
  }

  return _highlighter
}

const highlighterRef: {
  value?: Highlighter
} = {}

async function ensureHighlighter() {
  if (!highlighterRef.value)
    highlighterRef.value = await createHighlighterWithThemeAndLangs()
}

const highlighter: Highlighter = async (code, lang, meta) => {
  const cacheKey = JSON.stringify({ code, lang, meta })
  let cached = cache.get(cacheKey)
  if (cached)
    return cached
  await ensureHighlighter()
  const metaArray = (meta || '').split(' ')
  const containLineNumbers = metaArray.some(item => item.trim() === 'ln')
  const titleMeta = metaArray.find(item => item.startsWith('title='))
  const commandDoms: string[] = []
  const lines = code.split('\n')
  let noErrorsFirstLine: string | undefined
  if (lines[0] === '// @noErrors')
    noErrorsFirstLine = lines.shift()
  if (lang !== 'md') {
    code = lines.map((line, i) => {
      const [commandDomsInOneLine, newLine] = processCommands(line, i, lines.length)
      commandDoms.push(...commandDomsInOneLine)
      return newLine
    }).join('\n')
  }
  let title: string | undefined
  if (titleMeta)
    title = titleMeta.split('=')[1].replace(/(^")|("$)/g, '')
  if (noErrorsFirstLine)
    code = `${noErrorsFirstLine}\n${code}`

  cached = `
<div class="svp-code-block-wrapper">${title
? `<div class="svp-code-block--title">${title}</div>
`
: ''}
  <div class="svp-code-block${containLineNumbers ? ' svp-code-block--with-line-numbers' : ''}">
    ${commandDoms.join('\n')}
    <!-- svelte-ignore a11y-no-noninteractive-tabindex -->
    ${await highlighterRef.value?.(code, lang, meta)}
    <div class="svp-code-block--lang">
      ${lang}
    </div>
    <CopyCode />
    ${containLineNumbers ? `<div class="svp-code-block--line-numbers">${lines.map((_, i) => `<div class="svp-code-block--line-number-item">${i + 1}</div>`).join('\n')}</div>` : ''}
  </div>
</div>`

  cache.set(cacheKey, cached)
  return cached
}

export default highlighter
