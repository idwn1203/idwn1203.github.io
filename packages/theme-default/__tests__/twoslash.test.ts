import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { describe, expect, it } from 'vitest'

import highlighter from '../src/markdown/highlighter'

const svelteCode = readFileSync(resolve(import.meta.dirname, 'test.svelte'), 'utf-8')

describe('twoslash', () => {
  it('renderer floating svelte', async () => {
    const code = `
const num = 1

const foo = 'bar'

`

    expect(await highlighter(code, 'ts')).toMatchSnapshot()
  })

  it('twoslash svelte', async () => {
    expect(await highlighter(svelteCode, 'svelte')).toMatchSnapshot()
  })
})
