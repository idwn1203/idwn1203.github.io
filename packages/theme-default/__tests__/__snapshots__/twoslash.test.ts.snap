// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`twoslash > renderer floating svelte 1`] = `
"
<div class="svp-code-block-wrapper">
  <div class="svp-code-block">
    
    <!-- svelte-ignore a11y-no-noninteractive-tabindex -->
    <pre class="shiki shiki-themes vitesse-light night-owl" style="background-color:#ffffff;--shiki-dark-bg:#011627;color:#393a34;--shiki-dark:#d6deeb" tabindex="0"><code><span class="line"></span>
<span class="line"><span style="color:#AB5959;--shiki-dark:#C792EA">const</span><span style="color:#B07D48;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic"> num</span><span style="color:#999999;--shiki-dark:#C792EA"> =</span><span style="color:#2F798A;--shiki-dark:#F78C6C"> 1</span></span>
<span class="line"></span>
<span class="line"><span style="color:#AB5959;--shiki-dark:#C792EA">const</span><span style="color:#B07D48;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic"> foo</span><span style="color:#999999;--shiki-dark:#C792EA"> =</span><span style="color:#B5695999;--shiki-dark:#D9F5DD"> '</span><span style="color:#B56959;--shiki-dark:#ECC48D">bar</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">'</span></span>
<span class="line"></span>
<span class="line"></span></code></pre>
    <div class="svp-code-block--lang">
      ts
    </div>
    <CopyCode />
    
  </div>
</div>"
`;

exports[`twoslash > twoslash svelte 1`] = `
"
<div class="svp-code-block-wrapper">
  <div class="svp-code-block">
    
    <!-- svelte-ignore a11y-no-noninteractive-tabindex -->
    <pre class="shiki shiki-themes vitesse-light night-owl" style="background-color:#ffffff;--shiki-dark-bg:#011627;color:#393a34;--shiki-dark:#d6deeb" tabindex="0"><code><span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">&#x3C;</span><span style="color:#1E754F;--shiki-dark:#CAECE6">script</span><span style="color:#B07D48;--shiki-dark:#C5E478;font-style:inherit;--shiki-dark-font-style:italic"> context</span><span style="color:#999999;--shiki-dark:#7FDBCA">=</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">"</span><span style="color:#B56959;--shiki-dark:#ECC48D">module</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">"</span><span style="color:#B07D48;--shiki-dark:#C5E478;font-style:inherit;--shiki-dark-font-style:italic"> lang</span><span style="color:#999999;--shiki-dark:#7FDBCA">=</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">"</span><span style="color:#B56959;--shiki-dark:#ECC48D">ts</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">"</span><span style="color:#999999;--shiki-dark:#7FDBCA">></span></span>
<span class="line"><span style="color:#1E754F;--shiki-dark:#C792EA;font-style:inherit;--shiki-dark-font-style:italic">  export</span><span style="color:#AB5959;--shiki-dark:#C792EA"> const</span><span style="color:#B07D48;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic"> FOO</span><span style="color:#999999;--shiki-dark:#C792EA"> =</span><span style="color:#B5695999;--shiki-dark:#D9F5DD"> '</span><span style="color:#B56959;--shiki-dark:#ECC48D">BAR</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">'</span></span>
<span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">&#x3C;/</span><span style="color:#1E754F;--shiki-dark:#CAECE6">script</span><span style="color:#999999;--shiki-dark:#7FDBCA">></span></span>
<span class="line"></span>
<span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">&#x3C;</span><span style="color:#1E754F;--shiki-dark:#CAECE6">script</span><span style="color:#B07D48;--shiki-dark:#C5E478;font-style:inherit;--shiki-dark-font-style:italic"> lang</span><span style="color:#999999;--shiki-dark:#7FDBCA">=</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">"</span><span style="color:#B56959;--shiki-dark:#ECC48D">ts</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">"</span><span style="color:#999999;--shiki-dark:#7FDBCA">></span></span>
<span class="line"><span style="color:#1E754F;--shiki-dark:#C792EA;font-style:inherit;--shiki-dark-font-style:italic">  import</span><span style="color:#999999;--shiki-dark:#D6DEEB"> &#123;</span><span style="color:#B07D48;--shiki-dark:#D6DEEB"> onMount</span><span style="color:#999999;--shiki-dark:#D6DEEB"> &#125;</span><span style="color:#1E754F;--shiki-dark:#C792EA;font-style:inherit;--shiki-dark-font-style:italic"> from</span><span style="color:#B5695999;--shiki-dark:#D9F5DD"> '</span><span style="color:#B56959;--shiki-dark:#ECC48D">svelte</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">'</span></span>
<span class="line"></span>
<span class="line"><span style="color:#AB5959;--shiki-dark:#C792EA">  let</span><span style="color:#B07D48;--shiki-dark:#D6DEEB;font-style:inherit;--shiki-dark-font-style:italic"> count</span><span style="color:#999999;--shiki-dark:#C792EA"> =</span><span style="color:#2F798A;--shiki-dark:#F78C6C"> 0</span></span>
<span class="line"></span>
<span class="line"><span style="color:#59873A;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic">  onMount</span><span style="color:#999999;--shiki-dark:#D6DEEB">(</span><span style="color:#999999;--shiki-dark:#D9F5DD">()</span><span style="color:#999999;--shiki-dark:#C792EA"> =></span><span style="color:#999999;--shiki-dark:#D6DEEB"> &#123;</span></span>
<span class="line"><span style="color:#A0ADA0;--shiki-dark:#637777">    //</span><span style="color:#A0ADA0;--shiki-dark:#637777;font-style:inherit;--shiki-dark-font-style:italic"> mounted</span></span>
<span class="line"><span style="color:#999999;--shiki-dark:#D6DEEB">  &#125;)</span></span>
<span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">&#x3C;/</span><span style="color:#1E754F;--shiki-dark:#CAECE6">script</span><span style="color:#999999;--shiki-dark:#7FDBCA">></span></span>
<span class="line"></span>
<span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">&#x3C;</span><span style="color:#1E754F;--shiki-dark:#CAECE6">button</span><span style="color:#1E754F;--shiki-dark:#C792EA;font-style:inherit;--shiki-dark-font-style:italic"> on</span><span style="color:#999999;--shiki-dark:#7FDBCA">:</span><span style="color:#2E8F82;--shiki-dark:#7FDBCA">click</span><span style="color:#999999;--shiki-dark:#7FDBCA">=</span><span style="color:#999999;--shiki-dark:#D3423E">&#123;</span><span style="color:#999999;--shiki-dark:#D9F5DD">()</span><span style="color:#999999;--shiki-dark:#C792EA"> =></span><span style="color:#B07D48;--shiki-dark:#D6DEEB"> count</span><span style="color:#AB5959;--shiki-dark:#C792EA">++</span><span style="color:#999999;--shiki-dark:#D3423E">&#125;</span><span style="color:#999999;--shiki-dark:#7FDBCA">></span></span>
<span class="line"><span style="color:#393A34;--shiki-dark:#D6DEEB">  Count is: </span><span style="color:#999999;--shiki-dark:#D3423E">&#123;</span><span style="color:#B07D48;--shiki-dark:#D6DEEB">count</span><span style="color:#999999;--shiki-dark:#D3423E">&#125;</span></span>
<span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">&#x3C;/</span><span style="color:#1E754F;--shiki-dark:#CAECE6">button</span><span style="color:#999999;--shiki-dark:#7FDBCA">></span></span>
<span class="line"></span></code></pre>
    <div class="svp-code-block--lang">
      svelte
    </div>
    <CopyCode />
    
  </div>
</div>"
`;
